pkg/restore/restore.go

```diff
@@ -474,14 +485,55 @@ func (ctx *context) restoreFromDir(dir string) (api.RestoreResult, api.RestoreRe
 			if !existingNamespaces.Has(mappedNsName) {
 				logger := ctx.log.WithField("namespace", nsName)
 				ns := getNamespace(logger, filepath.Join(dir, api.ResourcesDir, "namespaces", api.ClusterScopedDir, nsName+".json"), mappedNsName)
-				if _, err := kube.EnsureNamespaceExists(ns, ctx.namespaceClient); err != nil {
+
+				clusterNS, err := ctx.namespaceClient.Get(mappedNsName, metav1.GetOptions{})
+				if err != nil && !apierrors.IsNotFound(err) {
 					addArkError(&errs, err)
 					continue
+				} else if apierrors.IsNotFound(err) {
+					_, err := ctx.namespaceClient.Create(ns)
+					if err != nil {
+						addArkError(&errs, err)
+						continue
+					}
+				}
+
+				// Namespace exists and is active
+				if clusterNS.Status.Phase == v1.NamespaceActive {
+					// keep track of namespaces that we know exist so we don't
+					// have to try to create them multiple times
+					existingNamespaces.Insert(mappedNsName)
+				}
+
+				// Namespace exists but is terminating
+				if clusterNS.Status.Phase == v1.NamespaceTerminating {
+					timeout := time.After(ctx.resourceTerminatingTimeout)
+					ticker := time.NewTicker(time.Second)
+					defer ticker.Stop()
+				loop:
+					for {
+						select {
+						// TODO: Need a way to break on process exit here; ctrl-c isn't working when we're in this loop
+						case <-timeout:
+							err := errors.Errorf("Timed out waiting on namespace %s to be ready", mappedNsName)
+							addArkError(&errs, err)
+							return warnings, errs
+						case <-ticker.C:
+							c, err := ctx.namespaceClient.Create(ns)
+							if apierrors.IsAlreadyExists(err) && c.Status.Phase == v1.NamespaceTerminating {
+								logger.Debugf("Namespace %s still terminating, waiting", mappedNsName)
+							} else if err != nil && !apierrors.IsAlreadyExists(err) {
+								addArkError(&warnings, err)
```

Comment:
if you move this block to a helper, then you can just return this error

Reply:
It can be returned, but it's going to warnings; would you suggest all the errors returned by the helper go into `&errs`?

Reply:
hmm, i think it makes sense to record this as an error, so yeah